src/blas/impl/KokkosBlas_Newton_impl.hpp

```diff
@@ -0,0 +1,250 @@
+  typename NormViewType::HostMirror lastResidualHost;
+
+  // NormViewType  residual_norms;
+  // TODO: Making these public for now. Should make private and access
+  // via setters and getters?
+  int maxIters;           // Maximum number of Newton steps
+  norm_type relativeTol;  // Relative convergence tolerance
+  bool debug_mode;        // Returns extra verbose output if true.
+
+  NewtonHandle(int _maxIters = 25, double _relativeTol = 1.0e-6,
+               bool _debug = false)
+      : lastResidual("ending Residual norm", 1),
+        lastResidualHost("end res norm host", 1),
+        maxIters(_maxIters),
+        relativeTol(_relativeTol),
+        debug_mode(_debug) {}
+
+  norm_type get_end_residual() const {
+    Kokkos::deep_copy(lastResidualHost, lastResidual);
+    return lastResidualHost(0);
+  }
+
+};  // NewtonHandle
+
+/// \brief Newton Functor:
+/// Solves the nonlinear system F(x) = 0
+/// where F is a map from R^n to R^m.
+/// \tparam LHSFunc: Functor on (J,x) where matrix J is output with
+/// the value of the Jacobian evaluated at x.
+/// \tparam RHSFunc: Functor on (y,x) where y is output with the value
+/// of F at x.
+/// \tparam Matrix: rank-3 view-type of LHSFunc output (extent(0) = 1)
+/// \tparam XVector: rank-2 view-type (extent(0) = 1)
+/// \tparam YVector: rank-2 view-type (extent(0) = 1)
+/// \param
+/// \param X [in]: Input vector X, a rank 2 view
+/// \param Y [in/out]: Output vector Y, a rank 2 view
+///
+/// No nested parallel_for is used inside of the function.
+///
+template <class System, class Matrix, class XVector, class YVector,
+          class NewtonHandleType>
+struct NewtonFunctor {
+  using execution_space = typename YVector::execution_space;
+  using yvalue_type     = typename YVector::non_const_value_type;
+  using norm_type       = typename NewtonHandleType::norm_type;
+
+  System sys;
+  XVector x;
+  YVector rhs;
+  NewtonHandleType handle;
+  int N;
+
+  Matrix J, tmp;
+  XVector update;
+  Kokkos::View<norm_type, execution_space> norm;
+  Kokkos::View<yvalue_type*, execution_space> alpha;
+
+  NewtonFunctor(System _sys, XVector _x, YVector _rhs,
+                NewtonHandleType& _handle)
+      : sys(_sys), x(_x), rhs(_rhs), handle(_handle) {
+    J      = Matrix("Jacobian", x.extent(0), x.extent(0));
+    tmp    = Matrix("Jacobian", x.extent(0), x.extent(0) + 4);
+    update = XVector("update", x.extent(0));
+    norm   = Kokkos::View<norm_type, execution_space>("Newton norm");
+    alpha  = Kokkos::View<yvalue_type*, execution_space>("alpha", 1);
+    N      = x.extent(0);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  NewtonSolverStatus solve() const {
+    alpha(0)               = Kokkos::ArithTraits<yvalue_type>::one();
+    handle.lastResidual(0) = -1;  // init to dummy value
+
+    // Iterate until maxIts or the tolerance is reached
+    for (int it = 0; it < handle.maxIters; ++it) {
+      // compute initial rhs
+      sys.derivatives(0, x, rhs);
```

Comment:
What is the purpose of the first argument and why is it always `0` and not used?
The `derivatives` name is a bit ambiguous to me, would `residual` be better?

Reply:
That's actually a convention from the ODE solver that allow the user to pass in a time `t` at which the functions are evaluated. This can be removed in favor of having the `sys` object handle that internally.

Reply:
That is what I was expecting but then should there be a way to pass the time `t`? It should not always be zero, it should be the value of the time `t` at the current time step; isn't it?